src/implementations/junoswap/pool.rs

```diff
@@ -0,0 +1,285 @@
+use cosmwasm_schema::cw_serde;
+use cosmwasm_std::{
+    to_binary, Addr, CosmosMsg, Decimal, Deps, Event, QuerierWrapper, QueryRequest, Response,
+    StdError, StdResult, Uint128, WasmMsg, WasmQuery,
+};
+use cw_asset::{Asset, AssetInfo, AssetList};
+use wasmswap::msg::{
+    ExecuteMsg, InfoResponse, QueryMsg, Token1ForToken2PriceResponse, TokenSelect,
+};
+
+use crate::{traits::Pool, CwDexError};
+
+use super::helpers::{
+    juno_get_lp_token_amount_to_mint, juno_get_token2_amount_required, JunoAsset, JunoAssetInfo,
+    JunoAssetList,
+};
+
+#[cw_serde]
+pub struct JunoswapPool {
+    pub addr: Addr,
+}
+
+impl JunoswapPool {
+    pub fn query_info(&self, querier: &QuerierWrapper) -> StdResult<InfoResponse> {
+        querier.query(&QueryRequest::Wasm(WasmQuery::Smart {
+            contract_addr: self.addr.to_string(),
+            msg: to_binary(&QueryMsg::Info {})?,
+        }))
+    }
+}
+
+impl Pool for JunoswapPool {
+    fn provide_liquidity(
+        &self,
+        deps: Deps,
+        assets: AssetList,
+        recipient: Addr,
+        slippage_tolerance: Option<Decimal>,
+    ) -> Result<Response, CwDexError> {
+        let pool_info = self.query_info(&deps.querier)?;
+
+        // Get the assets in the pool from the assets sent in
+        let juno_assets: JunoAssetList = assets.try_into()?;
+        let token1 = juno_assets.find(pool_info.token1_denom.into())?;
+        let token2 = juno_assets.find(pool_info.token2_denom.into())?;
+
+        // Calculate minimum LPs from slippage tolerance
+        let expected_lps = juno_get_lp_token_amount_to_mint(
+            token1.amount,
+            pool_info.lp_token_supply,
+            pool_info.token1_reserve,
+        )?;
+
+        // TODO: checked mul?
+        let min_liquidity =
+            expected_lps * Decimal::one().checked_sub(slippage_tolerance.unwrap_or_default())?;
+
+        let provide_liquidity = CosmosMsg::Wasm(WasmMsg::Execute {
+            contract_addr: self.addr.to_string(),
+            funds: vec![],
+            msg: to_binary(&ExecuteMsg::AddLiquidity {
+                token1_amount: token1.amount,
+                min_liquidity,
+                max_token2: token2.amount, // TODO: correct?
+                expiration: None,
+            })?,
```

Comment:
This will fail if the caller did not pass in enough of token2 to match the amount of token1, which is not the behavior we want. We want to maximize the amount of LP tokens returned and never have this fail, unless `slippage_tolerance` is crossed.